azurerm/resource_arm_log_profile.go

```diff
@@ -0,0 +1,259 @@
+					DiffSuppressFunc: azureRMSuppressLocationDiff,
+				},
+			},
+			"categories": {
+				Type:     schema.TypeList,
+				Required: true,
+				MinItems: 1,
+				Elem: &schema.Schema{
+					Type:             schema.TypeString,
+					DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+				},
+			},
+			"retention_policy": {
+				Type:     schema.TypeList,
+				Required: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"enabled": {
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+						"days": {
+							Type:     schema.TypeInt,
+							Required: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceArmLogProfileCreateOrUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).logProfilesClient
+	ctx := meta.(*ArmClient).StopContext
+
+	categories := expandLogProfileCategories(d)
+	locations := expandLogProfileLocations(d)
+	retentionPolicy, err := expandAzureRmLogProfileRetentionPolicy(d)
+	if err != nil {
+		return err
+	}
+
+	logProfileProperties := insights.LogProfileProperties{
+		Categories:      &categories,
+		Locations:       &locations,
+		RetentionPolicy: retentionPolicy,
+	}
+
+	storageAccountID := d.Get("storage_account_id").(string)
+	if storageAccountID != "" {
+		logProfileProperties.StorageAccountID = utils.String(storageAccountID)
+	}
+
+	serviceBusRuleID := d.Get("service_bus_rule_id").(string)
+	if serviceBusRuleID != "" {
+		logProfileProperties.ServiceBusRuleID = utils.String(serviceBusRuleID)
+	}
+
+	name := d.Get("name").(string)
+	parameters := insights.LogProfileResource{
+		Name:                 utils.String(name),
+		LogProfileProperties: &logProfileProperties,
+	}
+
+	_, createErr := client.CreateOrUpdate(ctx, name, parameters)
+	if createErr != nil {
+		return fmt.Errorf("Error Creating/Updating Log Profile %q: %+v", name, createErr)
+	}
+
+	// Wait for Log Profile to become available
+	err = resource.Retry(300*time.Second, retryLogProfilesClientGet(name, meta))
+	if err != nil {
+		return err
+	}
+
+	read, err := client.Get(ctx, name)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*read.ID)
+
+	return resourceArmLogProfileRead(d, meta)
+}
+
+func resourceArmLogProfileRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).logProfilesClient
+	ctx := meta.(*ArmClient).StopContext
+
+	name, err := parseLogProfileNameFromID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Get(ctx, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error making Read request on Azure Log Profile %q: %+v", name, err)
+	}
+
+	d.Set("name", resp.Name)
+	if props := resp.LogProfileProperties; props != nil {
+		d.Set("storage_account_id", props.StorageAccountID)
+		d.Set("service_bus_rule_id", props.ServiceBusRuleID)
+
+		locations := make([]string, len(*props.Locations))
```

Comment:
Although defined as required in Azure REST spec, we might still need to `nil` check `props.Locations` to prevent potential crash.

```go
if props.Locations != nil {
  locations := …
  …
  d.Set("locations", locations)
}
```